embassy-stm32/src/crc/v2.rs

```diff
@@ -0,0 +1,175 @@
+use crate::pac::crc::vals;
+use crate::pac::CRC as PAC_CRC;
+use crate::peripherals::CRC;
+use crate::rcc::sealed::RccPeripheral;
+
+pub struct Crc {
+    _peripheral: CRC,
+    _config: CrcConfig,
+}
+
+pub enum CrcConfigError {
+    InvalidPolynomial,
+}
+
+pub struct CrcConfig {
+    reverse_in: CrcInputReverseConfig,
+    reverse_out: bool,
+    #[cfg(crc_v3)]
+    poly_size: PolySize,
+    crc_init_value: u32,
+    #[cfg(crc_v3)]
+    crc_poly: u32,
+}
+
+pub enum CrcInputReverseConfig {
+    None,
+    Byte,
+    Halfword,
+    Word,
+}
+
+impl CrcConfig {
+    pub fn new(
+        reverse_in: CrcInputReverseConfig,
+        reverse_out: bool,
+        #[cfg(crc_v3)] poly_size: PolySize,
+        crc_init_value: u32,
+        #[cfg(crc_v3)] crc_poly: u32,
+    ) -> Result<Self, CrcConfigError> {
+        // As Per RM0091 (DocID018940 Rev 9), Even polynomials are not supported.
+        #[cfg(crc_v3)]
+        if crc_poly % 2 == 0 {
+            return Err(CrcConfigError::InvalidPolynomial);
+        }
+        Ok(CrcConfig {
+            reverse_in,
+            reverse_out,
+            #[cfg(crc_v3)]
+            poly_size,
+            crc_init_value,
+            #[cfg(crc_v3)]
+            crc_poly,
+        })
+    }
+}
+
+#[cfg(crc_v3)]
+pub enum PolySize {
+    Width7,
+    Width8,
+    Width16,
+    Width32,
+}
+
+impl Crc {
+    /// Instantiates the CRC32 peripheral and initializes it to default values.
+    pub fn new(peripheral: CRC, config: CrcConfig) -> Self {
+        // Note: enable and reset come from RccPeripheral.
+        // enable CRC clock in RCC.
+        CRC::enable();
+        // Reset CRC to default values.
+        CRC::reset();
+        let mut instance = Self {
+            _peripheral: peripheral,
+            _config: config,
+        };
+        CRC::reset();
+        instance.reconfigure();
+        instance.reset();
+        instance
+    }
+
+    pub fn reset(&mut self) {
+        unsafe {
+            PAC_CRC.cr().modify(|w| w.set_reset(true));
+        }
+    }
+
+    /// Reconfigures the CRC peripheral. Doesn't reset.
+    fn reconfigure(&mut self) {
+        unsafe {
+            // Init CRC value
+            PAC_CRC.init().write_value(self._config.crc_init_value);
+            #[cfg(crc_v3)]
+            PAC_CRC.pol().write_value(self._config.crc_poly);
+
+            // configure CR components
+            // (reverse I/O, polysize, poly)
+            PAC_CRC.cr().modify(|w| {
```

Comment:
This can be a `write` since you're setting all the fields. Maybe you can also reset it in the same write?

Reply:
I do not wish to write to reset here, since I don't know in what order the writes will occur.
Reset must be written to last.